File: fitness_core/potential.py
"""Maximum natural muscular potential (men). Returns max fat-free mass (kg);
pair with weight_at_bf to get max bodyweight at a target body-fat %.

Sources: Casey Butt (wrist+ankle FFM model); Kouri et al. natural FFMI ~25 cap;
Martin Berkhan max-contest-weight heuristic.
"""
import math

from fitness_core.units import LB_TO_KG

CM_TO_IN = 1 / 2.54


def weight_at_bf(ffm_kg: float, target_bf_pct: float) -> float:
    return ffm_kg / (1 - target_bf_pct / 100)


def casey_butt_ffm_kg(height_cm: float, wrist_cm: float, ankle_cm: float,
                      target_bf_pct: float) -> float:
    h = height_cm * CM_TO_IN
    w = wrist_cm * CM_TO_IN
    a = ankle_cm * CM_TO_IN
    lbm_lb = (h ** 1.5
              * (math.sqrt(w) / 22.6670 + math.sqrt(a) / 17.0104)
              * (target_bf_pct / 224 + 1))
    return lbm_lb * LB_TO_KG


def ffmi_cap_ffm_kg(height_cm: float) -> float:
    return 25 * (height_cm / 100) ** 2


def berkhan_ffm_kg(height_cm: float) -> float:
    return (height_cm - 100) * (1 - 0.055)

File: packages/core/src/math/potential.ts
import { FFMI_NATURAL_CAP } from "./indices.js";
import { LB_TO_KG } from "./units.js";

const CM_TO_IN = 1 / 2.54;

/** Total weight (kg) at a target body-fat % given fat-free mass (kg). */
export function weightAtBf(ffmKg: number, targetBfPct: number): number {
  return ffmKg / (1 - targetBfPct / 100);
}

/** Casey Butt's maximum drug-free fat-free mass (kg) from height/wrist/ankle (cm) and target body-fat %. */
export function caseyButtFfmKg(
  heightCm: number,
  wristCm: number,
  ankleCm: number,
  targetBfPct: number,
): number {
  const h = heightCm * CM_TO_IN;
  const w = wristCm * CM_TO_IN;
  const a = ankleCm * CM_TO_IN;
  const lbmLb =
    h ** 1.5 *
    (Math.sqrt(w) / 22.667 + Math.sqrt(a) / 17.0104) *
    (targetBfPct / 224 + 1);
  return lbmLb * LB_TO_KG;
}

/** Natural fat-free mass (kg) ceiling implied by an FFMI of 25, from height (cm). */
export function ffmiCapFfmKg(heightCm: number): number {
  return FFMI_NATURAL_CAP * (heightCm / 100) ** 2;
}

/** Martin Berkhan's "contest-shape" fat-free mass (kg) heuristic from height (cm). */
export function berkhanFfmKg(heightCm: number): number {
  return (heightCm - 100) * (1 - 0.055);
}
